### ISSUE TITLE
qt5-qmake-arm-linux-gnueabihf does not point to a Qt installation

### ISSUE THREAD

**Issue by mardy** [ubports/qtbase-opensource-src-packaging#2]
Hi! The `qt5-qmake-arm-linux-gnueabihf` binary, which we use in `crossbuilder` (and `clickable`, I suspect) does not carry the information about the Qt installation:

```
$ qt5-qmake-arm-linux-gnueabihf -query
QT_SYSROOT:
QT_INSTALL_PREFIX:
QT_INSTALL_ARCHDATA:
QT_INSTALL_DATA:
QT_INSTALL_DOCS:
QT_INSTALL_HEADERS:
QT_INSTALL_LIBS:
QT_INSTALL_LIBEXECS:
QT_INSTALL_BINS:
QT_INSTALL_TESTS:
QT_INSTALL_PLUGINS:
QT_INSTALL_IMPORTS:
QT_INSTALL_QML:
QT_INSTALL_TRANSLATIONS:
QT_INSTALL_CONFIGURATION:
QT_INSTALL_EXAMPLES:
QT_INSTALL_DEMOS:
QT_HOST_PREFIX:
QT_HOST_DATA:
QT_HOST_BINS:
QT_HOST_LIBS:
QMAKE_SPEC:
QMAKE_XSPEC:
QMAKE_VERSION:3.1
QT_VERSION:5.9.5
```

If I install an older version (`apt install qt5-qmake-arm-linux-gnueabihf=5.5.1+dfsg-16ubuntu7.5` in crossbuilder), then all the variables are properly set:
```
$ qt5-qmake-arm-linux-gnueabihf -query
QT_SYSROOT:
QT_INSTALL_PREFIX:/usr
QT_INSTALL_ARCHDATA:/usr/lib/arm-linux-gnueabihf/qt5
QT_INSTALL_DATA:/usr/share/qt5
QT_INSTALL_DOCS:/usr/share/qt5/doc
QT_INSTALL_HEADERS:/usr/include/arm-linux-gnueabihf/qt5
QT_INSTALL_LIBS:/usr/lib/arm-linux-gnueabihf
QT_INSTALL_LIBEXECS:/usr/lib/arm-linux-gnueabihf/qt5/libexec
QT_INSTALL_BINS:/usr/lib/x86_64-linux-gnu/qt5/bin
QT_INSTALL_TESTS:/usr/tests
QT_INSTALL_PLUGINS:/usr/lib/arm-linux-gnueabihf/qt5/plugins
QT_INSTALL_IMPORTS:/usr/lib/arm-linux-gnueabihf/qt5/imports
QT_INSTALL_QML:/usr/lib/arm-linux-gnueabihf/qt5/qml
QT_INSTALL_TRANSLATIONS:/usr/share/qt5/translations
QT_INSTALL_CONFIGURATION:/etc/xdg
QT_INSTALL_EXAMPLES:/usr/lib/arm-linux-gnueabihf/qt5/examples
QT_INSTALL_DEMOS:/usr/lib/arm-linux-gnueabihf/qt5/examples
QT_HOST_PREFIX:/usr
QT_HOST_DATA:/usr/lib/arm-linux-gnueabihf/qt5
QT_HOST_BINS:/usr/lib/x86_64-linux-gnu/qt5/bin
QT_HOST_LIBS:/usr/lib/arm-linux-gnueabihf
QMAKE_SPEC:linux-g++
QMAKE_XSPEC:ubuntu-arm-gnueabihf-g++
QMAKE_VERSION:3.0
QT_VERSION:5.5.1
```

Something must have been broken with the upgrade to Qt 5.9 (older version of the package, made for Qt 5.9, are affected by this bug as well).

**Comment by mardy** [comment]
It looks like the generation of the `qt5-qmake-arm-linux-gnueabihf` binary did, in fact, fail in the [last build](https://ci.ubports.com/job/qtbase-opensource-src-packaging/job/xenial/17/consoleFull):
```
[Build binary - amd64] Info: creating cache file /build/varlibjenkinsworkspacesourcesrcpackagingxenialWLENQADXJSSKBMTELKARDJTOZTMCYMZKFXRIEJERAWHSXQqtbaseopensourcesrcdfsgubportsdsc/.qmake.cache
[Build binary - amd64] Command line: -confirm-license -prefix /usr -bindir /usr/lib/x86_64-linux-gnu/qt5/bin -libdir /usr/lib/arm-linux-gnueabihf -docdir /usr/share/qt5/doc -headerdir /usr/include/arm-linux-gnueabihf/qt5 -datadir /usr/share/qt5 -archdatadir /usr/lib/arm-linux-gnueabihf/qt5 -hostdatadir /usr/share/qt5 -plugindir /usr/lib/arm-linux-gnueabihf/qt5/plugins -importdir /usr/lib/arm-linux-gnueabihf/qt5/imports -translationdir /usr/share/qt5/translations -hostdatadir /usr/lib/arm-linux-gnueabihf/qt5 -sysconfdir /etc/xdg -examplesdir /usr/lib/arm-linux-gnueabihf/qt5/examples -opensource --no-dbus --no-libjpeg --no-glib -platform linux-g++ -qt-zlib -no-rpath -verbose -optimized-qmake -no-strip -no-separate-debug-info -qpa xcb -xcb -no-icu -accessibility -no-directfb -no-sql-ibase -no-pch -no-opengl -xplatform ubuntu-arm-gnueabihf-g++
[Build binary - amd64] Project ERROR: Cannot run target compiler 'arm-linux-gnueabihf-g++'. Output:
[Build binary - amd64] ===================
[Build binary - amd64] sh: 1: arm-linux-gnueabihf-g++: not found
[Build binary - amd64] ===================
[Build binary - amd64] Maybe you forgot to setup the environment?
```

**Comment by mardy** [comment]
The error I quote in my previous comment is irrelevant (and actually, expected): the trick we have in the `debian/rules` file doesn't care about the whole configuration step, but only cares that `bin/qmake` gets generated (and that's why there's a ` || file $(CURDIR)/bin/qmake` at the end).

I'm still investigating, but I suspect that something was changed between Qt 5.5 and 5.9 which makes it so that the `qmake` binary generated at the configure step does not carry all the installation information within itself (maybe it relies on separate config files?). It might be that the [workaround](https://github.com/bhdouglass/clickable/blob/master/docker/ubuntu-sdk/16.04-armhf/qt.conf) adoped in `clickable` is actually the most sane way of solving this.

**Comment by fredldotme** [comment]
Any progress on this? Building packages with dpkg-buildpackage in an LXD container (using the SDK image) currently fails due to this error.

**Comment by mardy** [comment]
@fredldotme, what does `ls /usr/lib/*/qt5/*/bin/qt.conf` say?

Anyway, why don't you use [crossbuilder](https://github.com/ubports/crossbuilder)?

**Comment by fredldotme** [comment]
> @fredldotme, what does `ls /usr/lib/*/qt5/*/bin/qt.conf` say?

```
root@ubports-16-04:~# ls /usr/lib/*/qt5/*/bin/qt.conf
/usr/lib/x86_64-linux-gnu/qt5/arm-linux-gnueabihf/bin/qt.conf
```

> Anyway, why don't you use [crossbuilder](https://github.com/ubports/crossbuilder)?
Crossbuilder doesn't work with the current lxd snap package so I'm using a painfully-slow qemu-emulated armhf rootfs.

**Comment by mardy** [comment]
> ```
> root@ubports-16-04:~# ls /usr/lib/*/qt5/*/bin/qt.conf
> /usr/lib/x86_64-linux-gnu/qt5/arm-linux-gnueabihf/bin/qt.conf
> ```

This seems correct; what are its contents? And do you have a `qmake` in the same directory? If not, you probably should create it as a symlink, the way that crossbuilder does it: https://github.com/ubports/crossbuilder/blob/master/crossbuilder#L413

**Comment by dobey** [comment]
Just came across this when looking for things using this qmake and the ubuntu profile, as we try to get things working with clickable and Qt 5.12 for the upgrade…

Is this still relevant? Does anything actually still use it or the `ubuntu-arm-gnueabihf-g++` mkspec? I think it would be best to drop them with the 5.12 upgrade and things should use the `linux-arm-gnueabi-g++` mkspec instead, assuming that works.

**Comment by mardy** [comment]
Please see the lines https://github.com/ubports/crossbuilder/blob/75c55aa6c40c59c65e164066c7830d573c4532d2/crossbuilder#L564_L575

I think that, no matter how wrong the qmake configuration is, it can be easily worked around. I actually don't know if the latest version of the `qtbase-opensource-src` package still requires these overrides, but I guess it's not important.

As far as I'm concerned, this issue can be closed.

**Comment by dobey** [comment]
The `qt5-qmake-arm-gnueabihf` package has not been patched back in with the qt 5.12 upgrade, and I'd prefer to avoid it if possible (and it seems we can). Given your statements, I'll go ahead and close this. Thanks!
